documentation/source/coroutines.rst

```diff
@@ -202,3 +143,60 @@ type of generator function that can be iterated with ``async for``:
 
 More details on this type of generator can be found in :pep:`525`.
 
+
+.. _yield-syntax:
+
+Generator-based coroutines
+==========================
+
+.. note:: This style is no longer recommended and support may someday be removed.
+
+Prior to Python 3.5, and the introduction of :keyword:`async` and :keyword:`await`, coroutines were implemented as wrappers around generators.
+Coroutine functions would be decorated with :class:`~cocotb.decorators.coroutine` and would use :keyword:`yield` to block on other coroutines or triggers.
+
+.. code-block:: python3
+
+    @cocotb.coroutine
+    def wait_10ns():
+        cocotb.log.info("About to wait for 10 ns")
+        yield Timer(10, units='ns')
+        cocotb.log.info("Simulation time has advanced by 10 ns")
+
+    @cocotb.coroutine
+    def wait_100ns():
+        for _ in range(10):
+            yield wait_10ns()
+
+You may see existing code that uses this syntax for coroutines, but do not worry, it is compatible with async coroutines.
+Any place that will accept async coroutines, will also accept generator-based coroutines;
+including :func:`~cocotb.fork`.
+Likewise, any object that can be used in an :keyword:`await` statement can also be used in a :keyword:`yield` statement while in a generator-based coroutine;
+including triggers like :class:`~cocotb.triggers.Timer`.
+
+.. code-block:: python3
+
+    async def wait_100ns():
+        for _ in range(10):
+            await wait_10ns()
+
+.. code-block:: python3
+
+    @cocotb.coroutine
+    def clock(signal, half_period, half_period_units):
+        signal <= 0
+        while True:
+            yield Timer(half_period, half_period_units)
+            signal <= ~signal
+
+    @cocotb.test()
```

Comment:
Perhaps worth mentioning that `test` objects can be generators too without the coroutine decorator.